cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -180,12 +180,16 @@ private class ArrayContent extends Content, TArrayContent {
   override Type getType() { none() }
 }
 
-/**
- * Holds if data can flow from `node1` to `node2` via an assignment to `f`.
- * Thus, `node2` references an object with a field `f` that contains the
- * value of `node1`.
- */
-predicate storeStep(Node node1, Content f, PostUpdateNode node2) {
+private predicate storeStepNoChi(Node node1, Content f, PostUpdateNode node2) {
+  exists(FieldAddressInstruction fa, StoreInstruction store |
+    store = node2.asInstruction() and
+    store.getDestinationAddress() = fa and
+    store.getSourceValue() = node1.asInstruction() and
+    f.(FieldContent).getField() = fa.getField()
```

Comment:
How do you avoid overlap between `storeStepNoChi` and `storeStepChi`? Won't the former match every store that the latter matches?

Reply:
It won't overlap since there is no `PostUpdateNode` that is both a `ChiInstruction` and a `StoreInstruction`. In particular, in `storeStepNoChi` we have `store = node2.asInstruction() and` and in `storeStepChi` we have `node2.asInstruction().(ChiInstruction).getPartial() = store`.

Reply:
I see. It's enforced in the charpred of `ExplicitSingleFieldStoreQualifierNode`.